T47 support

niden opened the pull request:
Hello!

* Type: new feature
* Link to issue: #47 

**In raising this pull request, I confirm the following:**

- [x] I have read and understood the [Contributing Guidelines](https://github.com/phalcon/phalcon/blob/master/CONTRIBUTING.md)
- [x] I have checked that another pull request for this purpose does not exist
- [x] I wrote some tests for this PR
- [ ] I have updated the relevant CHANGELOG
- [ ] I have created a PR for the [documentation](https://github.com/phalcon/docs) about this change

Support namespace - refactoring the helper namespace

Thanks

niden committed:
refactor(helper): removed helper classes in favor of support namespace

Refs #47

niden committed:
refactor(str): refactored the str helper classes into smaller code fragments

Refs #47

niden committed:
refactor(tests): reworked the helper-str tests in the support namespace

Refs #47

niden committed:
refactor(helper): refactored json classes to the support namespace

Refs #47

niden committed:
refactor(helper): refactored helper/number to the support namespace

Refs #47

niden committed:
style: code sniffer changes

Refs #47

niden committed:
refactor(helper): arr related classes refactoring; moved each method to its own class

Refs #47

niden committed:
refactor(helper): refactored helper/fs class to work with the new support helper factory

Refs #47

niden committed:
refactor(helper): reformatting mostly as well as minor refactoring

Refs #47

niden committed:
refactor: renamed getAdapters to getServices

Refs #47

niden committed:
style: code sniffer corrections

Refs #47

niden committed:
test: corrected test assertion

Refs #47

niden committed:
feature(setup): removed codacy code reporter for now

Refs #47

ruudboon commented on src/Logger/AdapterFactory.php:
```diff
@@ -47,7 +37,7 @@ public function __construct(array $services = [])
      *
      * @return AdapterInterface
-     * @throws ExceptionAlias
+     * @throws SupportException
```
Will a namespace alias work with the doc's generation?

niden replied:
I believe it does

Jeckerson replied:
Yes it does

ruudboon commented on src/Support/Arr/Order.php:
```diff
@@ -0,0 +1,61 @@
+    public function __invoke(
+        array $collection,
+        $attribute,
+        string $order = 'asc'
```
Should we add sort flags as well? https://www.php.net/manual/en/function.sort.php#refsect1-function.sort-parameters
How about constants for the asc/desc?

niden replied:
Added sort flags in the method - same as the ones from PHP

ruudboon commented on src/Support/Arr/Order.php:
```diff
@@ -0,0 +1,61 @@
+            $sorted[$key] = $item;
+        }
+
+        if ('asc' === $order) {
```
If we stay with string I wouldn't mind allowing 'ASC' in capitals as well `==`

niden replied:
Added constants for this - no confusion or additional checks.

ruudboon commented on src/Support/Str/Friendly.php:
```diff
@@ -0,0 +1,96 @@
+        );
+
+        if ($lowercase) {
+            $friendly = $this->toLower((string) $friendly);
```
Since text is checked as string in parameters I think it's safe to handle the output of preg_replace as string and casting isn't needed.

niden replied:
Done

niden committed:
refactor: changes based on CR comments

Refs #47

niden committed:
test(codacy): trying a different syntax to upload the coverage report

Refs #47

niden committed:
refactor: trying a different way to set the codacy token

Refs #47

niden committed:
refactor: trying different format for the token

Refs #47

niden committed:
refactor: minor refactoring on flatten/group/friendly; removed codacy for now

Refs #47

niden committed:
refactor: json::encode flags; arr::order conditional; arr::set conditional

changed the exception thrown by json::encode to be JsonException. Added default options to be JSON_HEX_TAG, JSON_HEX_APOS, JSON_HEX_AMP, JSON_HEX_QUOT, JSON_UNESCAPED_SLASHES and JSON_THROW_ON_ERROR. The conditional still remains to throw an exception on error in case the user does not use that option for `json_encode`. Refactored Order and Set (Arr) to no longer use `else` statements

Refs #47

niden committed:
test: logic corrections because of tests !!

Refs #47

niden committed:
test(messages): added more messages tests for coverage

Refs #47

niden committed:
refactor: minor phpcs corrections

Refs #47

niden committed:
style: style adjustments

Refs #47

Jeckerson commented on src/Support/Arr/Chunk.php:
```diff
@@ -0,0 +1,41 @@
+/**
+ * This file is part of the Phalcon.
+ *
+ * (c) Phalcon Team <team@phalcon.com>
```
```suggestion
 * (c) Phalcon Team <team@phalcon.io>
```

Jeckerson commented on src/Support/Arr/Blacklist.php:
```diff
@@ -0,0 +1,55 @@
+/**
+ * This file is part of the Phalcon.
+ *
+ * (c) Phalcon Team <team@phalcon.com>
```
```suggestion
 * (c) Phalcon Team <team@phalcon.io>
```

niden replied:
Let me do a global replace in case I missed this elsewhere

niden replied:
Resolved with latest push

Jeckerson commented on src/Support/Arr/First.php:
```diff
@@ -0,0 +1,44 @@
+/**
+ * This file is part of the Phalcon.
+ *
+ * (c) Phalcon Team <team@phalcon.com>
```
```suggestion
 * (c) Phalcon Team <team@phalcon.io>
```

Jeckerson commented on src/Support/Arr/FirstKey.php:
```diff
@@ -0,0 +1,44 @@
+/**
+ * This file is part of the Phalcon.
+ *
+ * (c) Phalcon Team <team@phalcon.com>
```
```suggestion
 * (c) Phalcon Team <team@phalcon.io>
```

Jeckerson commented on src/Support/Arr/Flatten.php:
```diff
@@ -0,0 +1,95 @@
+/**
+ * This file is part of the Phalcon.
+ *
+ * (c) Phalcon Team <team@phalcon.com>
```
```suggestion
 * (c) Phalcon Team <team@phalcon.io>
```

Jeckerson commented on src/Support/Arr/Get.php:
```diff
@@ -0,0 +1,50 @@
+/**
+ * This file is part of the Phalcon.
+ *
+ * (c) Phalcon Team <team@phalcon.com>
```
```suggestion
 * (c) Phalcon Team <team@phalcon.io>
```

Jeckerson commented on src/Support/Arr/Get.php:
```diff
@@ -0,0 +1,50 @@
+    ) {
+        $value = $collection[$index] ?? $defaultValue;
+
+        if (is_string($cast)) {
```
```suggestion
        if ($cast !== null) {
```

faster then call function.

niden replied:
Resolved

niden committed:
style(docblock): corrected email address

Refs #47

Jeckerson commented on src/Support/Arr/Group.php:
```diff
@@ -0,0 +1,140 @@
+     *
+     * @return bool
+     */
+    private function isSet($method, $element): bool
```
exist?

niden replied:
Nah this is a private method; used to check the `isset` condition and in turn avoid using `else` statements

Jeckerson commented on src/Support/Arr/Pluck.php:
```diff
@@ -0,0 +1,47 @@
+        foreach ($collection as $item) {
+            if (is_object($item) && isset($item->{$element})) {
+                $filtered[] = $item->{$element};
+            } elseif (is_array($item) && isset($item[$element])) {
```
AFAIK, in both cases: object and array, would be enough to just use isset().

niden replied:
I am going to refactor this to remove the `else` statement but it needs to check about an object or an array because the assignment is not the same on both cases. Not sure if we can assume that this is going to be an array if `isset($a[b])` is true (thus removing the `is_array`)

niden committed:
perf(support-arr): introduced faster `if` statement

Refs #47

niden committed:
refactor(support---arr---pluck): removed else statement

Refs #47
